pkg/operator/controller/route-status/route_status_test.go

```diff
@@ -0,0 +1,469 @@
+			}
+		})
+	}
+}
+
+// newIngressControllerWithSelectors returns a new ingresscontroller with the specified name,
+// routeSelectors, and namespaceSelectors based on the parameters.
+func newIngressControllerWithSelectors(name, routeMatchLabel, routeMatchExpression, namespaceMatchLabel, namespaceMatchExpression string) *operatorv1.IngressController {
+	ingresscontroller := operatorv1.IngressController{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: operatorcontroller.DefaultOperatorNamespace,
+		},
+	}
+	if len(routeMatchLabel) != 0 {
+		ingresscontroller.Spec.RouteSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"type": routeMatchLabel,
+			},
+		}
+	}
+	if len(routeMatchExpression) != 0 {
+		ingresscontroller.Spec.RouteSelector = &metav1.LabelSelector{
+			MatchExpressions: []metav1.LabelSelectorRequirement{
+				{
+					Key:      "type",
+					Operator: metav1.LabelSelectorOpIn,
+					Values:   []string{routeMatchExpression},
+				},
+			},
+		}
+	}
+	if len(namespaceMatchLabel) != 0 {
+		ingresscontroller.Spec.NamespaceSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"type": namespaceMatchLabel,
+			},
+		}
+	}
+	if len(namespaceMatchExpression) != 0 {
+		ingresscontroller.Spec.NamespaceSelector = &metav1.LabelSelector{
+			MatchExpressions: []metav1.LabelSelectorRequirement{
+				{
+					Key:      "type",
+					Operator: metav1.LabelSelectorOpIn,
+					Values:   []string{namespaceMatchExpression},
+				},
+			},
+		}
+	}
+	return &ingresscontroller
+}
+
+// newNamespace returns a new namespace with the specified name
+// and if label exists, with that label
+func newNamespace(name, label string) *corev1.Namespace {
+	namespace := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+	}
+	if len(label) != 0 {
+		namespace.ObjectMeta.Labels = map[string]string{"type": label}
+	}
+	return &namespace
+}
+
+// newRouteWithAdmittedStatuses returns a new route that is admitted by ingress controllers.
+func newRouteWithLabelWithAdmittedStatuses(name, namespace, label string, icAdmitted ...string) *routev1.Route {
+	route := newRouteWithAdmittedStatuses(name, icAdmitted...)
+	route.ObjectMeta.Namespace = namespace
+	if len(label) != 0 {
+		route.Labels = map[string]string{
+			"type": label,
+		}
+	}
+	return route
+}
+
+// newRouteWithAdmittedStatuses returns a new route that is admitted by ingress controllers.
+func newRouteWithAdmittedStatuses(name string, icAdmitted ...string) *routev1.Route {
+	route := routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+	}
+	if len(icAdmitted) != 0 {
+		for _, ic := range icAdmitted {
+			route.Status.Ingress = append(route.Status.Ingress, routev1.RouteIngress{
+				RouterName: ic,
+				Conditions: []routev1.RouteIngressCondition{
+					{
+						Type:   routev1.RouteAdmitted,
+						Status: "True",
+					},
+				},
+			})
+		}
+	}
+	return &route
+}
```

Comment:
Why do these all return pointer values?  Is that just for the one test case in `Test_Reconcile` where `expectedRoute` is nil?  

I wonder how we could make the test code easier to read.  One option would be to use the builder pattern, like what Andrey did here: https://github.com/openshift/aws-load-balancer-operator/blob/54506d26e0de8cbc71289d897b71b39e1c185391/test/e2e/util.go#L274-L337

Reply:
**Update**: I added the builder pattern in https://github.com/openshift/cluster-ingress-operator/pull/849/commits/efc08f91fc67ffb060cb4616f78894171f08ccb7. It's reusable, and hopefully much more readable.